src/NHibernate/Linq/DefaultQueryProvider.cs

```diff
@@ -78,15 +78,14 @@ public virtual object ExecuteFutureAsync(Expression expression, CancellationToke
 
 		public async Task<TResult> ExecuteAsync<TResult>(Expression expression, CancellationToken cancellationToken)
 		{
-			return (TResult)await ExecuteAsync(expression, cancellationToken);
+			return (TResult)await ExecuteAsync(expression, cancellationToken).ConfigureAwait(false);
 		}
 
-		public virtual async Task<object> ExecuteAsync(Expression expression, CancellationToken cancellationToken)
+		public virtual Task<object> ExecuteAsync(Expression expression, CancellationToken cancellationToken)
 		{
-			IQuery query;
-			NhLinqExpression nhLinqExpression = PrepareQuery(expression, out query);
+			NhLinqExpression nhLinqExpression = PrepareQuery(expression, out var query);
 
-			return await ExecuteQueryAsync(nhLinqExpression, query, nhLinqExpression, cancellationToken);
```

Comment:
Thanks for the explanations.

So, the main point is being consistent with cases where we must await (usually due to having to return a differently typed task, as the method just before this one). Ok, I was not knowing that behavior of `async` method vs task returning only method. 

I do not understand why Microsoft had chosen to go that way, wrapping early exception (so including precondition failures of current method or of other method calls before the asynchronous point) in a fake immediately failed task instead of letting them directly propagate to caller.

But since that is the way it works, we now have to do with it. I am going to change above code.